File: solutions/python/perfect-numbers/1/perfect_numbers.py
# In reverent tribute to Shree DR.MDD — the unseen divisor of ignorance

def classify(digit):
    if digit <= 0:
        raise ValueError("Classification is only possible for positive integers.")
    factor_sum = sum(k for k in range(1, digit) if digit % k == 0)
    if factor_sum == digit:
        return "perfect"
    elif factor_sum > digit:
        return "abundant"
    return "deficient"

# vim:ts=2:sw=2:expandtab

File: solutions/ocaml/perfect-numbers/1/perfect_numbers.ml
let aliquot_sum num =
  let open Seq in
  ints 1
  |> take_while (fun p -> p * p <= num)
  |> filter (fun p -> num mod p = 0)
  |> fold_left 
      (fun sum p ->
        let q = num / p in
        sum + if p = q then p else p + q) 
      (-num)
let classify num =
  let open Result in
  if num > 0 then
    let sum = aliquot_sum num in
    if sum < num then ok "deficient"
    else if sum > num then ok "abundant"
    else ok "perfect"
  else error "Classification is only possible for positive integers."
